crypto/param_trie.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include "crypto/param_trie.h"
+#include "crypto/ctype.h"
+#include "internal/thread_once.h"
+
+static unsigned char alphabet_map[256];
+
+static void init_alphabet_table(void)
+{
+    unsigned int i;
+    unsigned char *a;
+#ifdef FIPS_MODULE
+    unsigned char aa[256];
+
+    a = aa;
+#else
+    a = alphabet_map;
+#endif
+
+    memset(a, 255, sizeof(alphabet_map));
+    for (i = 0; i < ALPHABET_SIZE; i++) {
+        a[(unsigned char)alphabet[i]] = i;
+        a[(unsigned char)ossl_tolower(alphabet[i])] = i;
+    }
+#ifdef FIPS_MODULE
+    memcpy(alphabet_map, a, sizeof(alphabet_map));
```

Comment:
Isn't this going to throw up tsan errors?

Reply:
Yes :(  It's harmless but still.

Without access to the run once support, I can't think of anything better.  An explicit lock seems like overkill.

I plan on adding small memory support which falls back to the OSSL_PARAM_locate() calls.  I could use this for FIPS mode as well which solves this concern.

Reply:
Actually, that seems like a reasonable solution.  I'll enforce the small memory mode when building for FIPS.

Reply:
Weird, I got build errors without the FIPS conditionals......

Reply:
Giving this a spin on the CIs.

Reply:
It needs the magic:
```c
#define ALLOW_RUN_ONCE_IN_FIPS
#include "internal/thread_once.h"
```